### ISSUE TITLE
Interpolation issue in point Initialization

### ISSUE THREAD

**Issue by santiagoMonedero** [firelab/windninja#418]
Windninja seems to do a wrong interpolation when the timeStamp of the weather stations is identical to the timeStamp of the time steps used in the simulation (when time interpolation is not necessary). I think it is all related with the calculation of `posMasterIdx `and `negMasterIdx`. For example, using 1 station with 5 measurements and 5 time steps with identical timeStamps,  I get the following vectors:
```
posMasterIdx =[-1, 0, 1, 2, 3]
negMasterIdx = [1, 2, 3, 4, -1]
```
which causes the wind to be interpolated in the following way 
```
vInter[0] = v[1]  //No interpolation, use closest negative step (future)
vInter[k] = (v[k-1]+v[k+1])/2  //interpolate between K-1 and K+1 (!?)
vInter[4] = v[3]  //No interpolation, use closest positive step (past)
```
Instead of `vInter[k] = v[k]` as would be expected. Here you can see the result with an actual station file

![image](https://user-images.githubusercontent.com/58716318/125216527-0abbaf00-e2f1-11eb-85d5-9a1102b126b3.png)

here is the csv I used:
```
"Station_Name","Coord_Sys(PROJCS,GEOGCS)","Datum(WGS84,NAD83,NAD27)","Lat/YCoord","Lon/XCoord","Height","Height_Units(meters,feet)","Speed","Speed_Units(mph,kph,mps,kts)","Direction(degrees)","Temperature","Temperature_Units(F,C)","Cloud_Cover(%)","Radius_of_Influence","Radius_of_Influence_Units(miles,feet,meters,km)","date_time"
0,GEOGCS,WGS84,38.7486675,-122.48,20,feet,5,kph,180,30,C,0,-1,km,2030-06-18T10:00:00Z
0,GEOGCS,WGS84,38.7486675,-122.48,20,feet,20,kph,270,30,C,0,-1,km,2030-06-18T11:00:00Z
0,GEOGCS,WGS84,38.7486675,-122.48,20,feet,30,kph,0,30,C,0,-1,km,2030-06-18T12:00:00Z
0,GEOGCS,WGS84,38.7486675,-122.48,20,feet,40,kph,95,30,C,0,-1,km,2030-06-18T13:00:00Z
0,GEOGCS,WGS84,38.7486675,-122.48,20,feet,50,kph,270,30,C,0,-1,km,2030-06-18T14:00:00Z
```

and the cfg:
```
num_threads=1
initialization_method=pointInitialization
match_points=true
output_wind_height=20
units_output_wind_height=ft
units_ascii_out_resolution=m
units_mesh_resolution=m
write_ascii_output=true
vegetation=trees
output_speed_units=kph
wx_station_filename=C:/WFAapps/FIRESIM/windninja/temp/stations.csv
elevation_file=C:/WFAapps/FIRESIM/windninja/temp/mdt.tif
number_time_steps=5
time_zone=UTC
mesh_resolution=300
ascii_out_resolution=300
start_year=2030
start_month=06
start_day=18
start_hour=10
start_minute=00
stop_year=2030
stop_month=06
stop_day=18
stop_hour=14
stop_minute=00
output_path=C:/WFAapps/FIRESIM/windninja/temp/
write_wx_station_csv=true
```

Thanks

**Comment by santiagoMonedero** [comment]
Hi @nwagenbrenner , I have modified the `interpolateTimeData `method to try to fix the interpolation problem. The algorithm is much more straight forward and concise than the previous one, so I am not sure If I may be missing something.  I have tested a few samples and it seems to be working as expected. In case you want to have a look at it, or even implement it if you see it robust enough, here is the code:

```
vector<vector<pointInitialization::preInterpolate> > pointInitialization::interpolateTimeData(std::string demFileName,
                        vector<vector<pointInitialization::preInterpolate> > vecStations,
                        std::vector<bpt::ptime> timeList)
{
    CPLDebug("STATION_FETCH","Interpolating Weather Data...");
  
    vector<vector<preInterpolate> > interpolatedWxData;
    for(int k=0; k< vecStations.size(); k++) 
    {
        vector<preInterpolate> subInter;
        for(int i=0; i<timeList.size(); i++) 
        {
            // We assume data is unordered. Find closest item in the past (idx0) and future (idx1). Find first (minIdx) and last (maxIdx) items. 
            //----------------------------------------------------------------------------------------------
            int idx0 = -1;  int idx1 = -1;
            int minIdx = 0; int maxIdx = 0;
            bpt::time_duration delta0= bpt::time_duration(-99999, 0, 0, 0);;
            bpt::time_duration delta1= bpt::time_duration(+99999, 0, 0, 0);;;
            bpt::time_duration zero(0, 0, 0, 0);

            for (int mm = 0; mm < vecStations[k].size(); mm++) 
            {
                bpt::time_duration delta = vecStations[k][mm].datetime - timeList[i];
                if (delta >= zero && delta <= delta1) { idx1 = mm; delta1 = delta; }
                if (delta <= zero && delta >= delta0) { idx0 = mm; delta0 = delta; }
                if (vecStations[k][mm].datetime < vecStations[k][minIdx].datetime) { minIdx = mm; }
                if (vecStations[k][mm].datetime > vecStations[k][maxIdx].datetime) { maxIdx = mm; }
            }

            if (idx0 == -1 && idx1 == -1) { throw std::runtime_error("User Provided start and stop times are both outside datasets time span!"); }
            if (idx0 == -1) { idx0 = minIdx; } 
            if (idx1 == -1) { idx1 = maxIdx; }

            //Interpolation weight w0
            //------------------------
            double t0 = unixTime(vecStations[k][idx0].datetime); 
            double t1 = unixTime(vecStations[k][idx1].datetime);
            double w0 = 1;
            if (t1 > t0) { w0 =(t1 - unixTime(timeList[i])) / (t1 - t0); }

            //Interpolate
            //---------------------
            preInterpolate interpol;
            interpol.datetime = timeList[i];
            interpol.speed = w0 * vecStations[k][idx0].speed + (1 - w0) * vecStations[k][idx1].speed ;
            interpol.temperature = w0 * vecStations[k][idx0].temperature + (1 - w0) * vecStations[k][idx1].temperature;
            interpol.cloudCover = w0 * vecStations[k][idx0].cloudCover + (1 - w0) * vecStations[k][idx1].cloudCover;
            double xx = w0 * sin(vecStations[k][idx0].direction * PI / 180.0) + (1 - w0) * sin(vecStations[k][idx1].direction * PI / 180.0) ;
            double yy = w0 * cos(vecStations[k][idx0].direction * PI / 180.0) + (1 - w0) * cos(vecStations[k][idx1].direction * PI / 180.0);
            double angle = atan2(xx, yy) * 180.0 / PI;
            if (angle < 0.0) { angle += 360.0; }
            interpol.direction = angle;

           // Static data
           //----------------------------
            interpol.lat = vecStations[k][0].lat;
            interpol.lon = vecStations[k][0].lon;
            interpol.datumType = vecStations[k][0].datumType;
            interpol.coordType = vecStations[k][0].coordType;
            interpol.height = vecStations[k][0].height;
            interpol.heightUnits = vecStations[k][0].heightUnits;
            interpol.influenceRadius = vecStations[k][0].influenceRadius;
            interpol.influenceRadiusUnits = vecStations[k][0].influenceRadiusUnits;
            interpol.stationName = vecStations[k][0].stationName;
            interpol.inputSpeedUnits = vecStations[k][0].inputSpeedUnits;
            interpol.tempUnits = vecStations[k][0].tempUnits;
            interpol.cloudCoverUnits = vecStations[k][0].cloudCoverUnits;

            subInter.push_back(interpol);
        }

        interpolatedWxData.push_back(subInter);
    }
    return interpolatedWxData;
}
```

**Comment by nwagenbrenner** [comment]
@santiagoMonedero Thanks for filing these issues and for this proposed fix. I'm in the field right now, but will take a look as soon as I can.

**Comment by nwagenbrenner** [comment]
@santiagoMonedero Can you submit your fix as a pull request? If it works, we'll merge it as a fix. Might be a week or so before I can do some testing though. Thanks!

**Comment by santiagoMonedero** [comment]
@nwagenbrenner I also noticed that the code assumes meters as the station height unit, so it gives  wrong results if the user sets the height in feets  (i.e. 20 feet is assumed to be 65 feets in the code !)
https://github.com/firelab/windninja/blob/master/src/ninja/pointInitialization.cpp#L1446

I think the pull request also solves this problem. I did not want to touch much, but if the fix works  the demFileName input argument, and the directTemporalInterpolation , interpolator, and interpolateDirection methods could be removed.
